### PR TITLE
osd: allow sparse reads with a non-zero truncate-seq

### PR DESCRIPTION
**Pull request by jtlayton**
do_read() just uses the truncate_seq to tell how to cap the length of
the read. I see no reason that sparse reads should do anything
differently.

Change do_sparse_read() to cap the requested length at the truncate_size
if the truncate_seq in the request is newer than we have in the object.

Fixes: https://tracker.ceph.com/issues/54280
Signed-off-by: Jeff Layton <jlayton@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by jtlayton** [comment]
Just a draft for now, as I'm still working on the kernel client pieces. I think we'll need something like this though.

@yehudasa : do you remember why sparse read support wasn't initially supported? It does look like it may just have been difficult to do with the initial code, but now that we have proper fiemap support I don't see why we shouldn't support this.

**Comment by gregsfortytwo** [comment]
This looks good to me as far as it goes. 👍 We just need to work out testing for it.

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -5846,21 +5846,23 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
...
     // translate sparse read to a normal one if not supported
     uint64_t offset = op.extent.offset;
     uint64_t length = op.extent.length;
-    if (offset > oi.size) {
+    if (offset > size) {
       length = 0;
-    } else if (offset + length > oi.size) {
-      length = oi.size - offset;
+    } else if (offset + length > size) {
+      length = size - offset;
```
Comment (lxbsz) [inline src/osd/PrimaryLogPG.cc:5865]: Should we also do something like this in none erasure case ?

Comment (jtlayton) [reply src/osd/PrimaryLogPG.cc:5865]: Maybe. I'm still seeing some data corruption errors with fsx, and that might be the culprit. I'll play with a revised patch today.

One thing I'm not clear on is that extent.length will end up being reset below around like 5920. Why is that being done?

**Comment by jtlayton** [comment]
The current patch works in my testing. I'm looking at how to add a testcase now.

**Comment by yehudasa** [comment]
@jtlayton iirc sparse read wasn't initially supported due to some filesystem bug that we encountered at the time (maybe in btrfs?) that returned bad results.

**Comment by jtlayton** [comment]
@yehudasa : ok, thanks. That makes some sense I think. I've rolled a small testcase for it too, so I think this is now ready for review/merge.

**Comment by jtlayton** [comment]
jenkins test make check

File: src/osd/PrimaryLogPG.cc (hunk 1 above)
Comment (jtlayton) [reply src/osd/PrimaryLogPG.cc]: Latest set limits the fiemap call to the truncate_size when truncate_seq is larger than what the MDS has.

**Review by lxbsz — APPROVED** [review]
This looks good to me.

File: src/osd/PrimaryLogPG.cc [hunk 2]
```diff
@@ -5846,22 +5846,24 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
...
   auto& soid = oi.soid;
+  uint64_t size = oi.size;
+  uint64_t offset = op.extent.offset;
+  uint64_t length = op.extent.length;
 
-  if (op.extent.truncate_seq) {
-    dout(0) << "sparse_read does not support truncation sequence " << dendl;
-    return -EINVAL;
-  }
+    // are we beyond truncate_size?
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:5853]: wrong indent.

File: src/osd/PrimaryLogPG.cc [hunk 3]
```diff
@@ -5846,22 +5846,24 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
...
+  if ( (oi.truncate_seq < op.extent.truncate_seq) &&
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:5854]: could you please remove the space between the `(`s?

File: src/osd/PrimaryLogPG.cc [hunk 4]
```diff
@@ -5846,22 +5846,24 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
...
+       (op.extent.offset + op.extent.length > op.extent.truncate_size) &&
+       (size > op.extent.truncate_size) )
+    size = op.extent.truncate_size;
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:5857]: could you add curly brackets around the single-line statement just for the sake of better readability?

File: src/osd/PrimaryLogPG.cc [hunk 5]
```diff
@@ -5846,22 +5846,24 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
...
+
+  if (offset > size)
+    length = 0;
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:5860]: ditto

File: src/osd/PrimaryLogPG.cc [hunk 6]
```diff
@@ -5846,22 +5846,24 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
...
+  else if (offset + length > size)
+    length = size - offset;
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:5862]: and ditto.

File: src/osd/PrimaryLogPG.cc [hunk 7]
```diff
@@ -5883,9 +5885,10 @@ int PrimaryLogPG::do_sparse_read(OpContext *ctx, OSDOp& osd_op) {
   } else {
     // read into a buffer
     map<uint64_t, uint64_t> m;
+
```
Comment (tchaikov) [inline src/osd/PrimaryLogPG.cc:5888]: i'd suggest remove the vertical space.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by jtlayton** [comment]
Thanks @tchaikov . I fixed up those nits and re-pushed.

**Comment by sseshasa** [comment]
### Teuthology Test Result

**Unrelated Failures**
1. JobIDs: [6739343](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739343/),  [6739544](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739544/)
- Test related to "workloads/admin_socket_objecter_requests" failed.
- New tracker created: https://tracker.ceph.com/issues/54604

2. JobID: [6739491](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739491/)
- Valgrind errors reported on mon.a
-  New tracker created: https://tracker.ceph.com/issues/54603

3. JobID:  [6739497](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739497/)
- Failure Reason: Unit test failure in ObjectStore/StoreTestSpecificAUSize.BluestoreBrokenNoSharedBlobRepairTest/2
- src/os/bluestore/BlueStore.cc: 7685: FAILED ceph_assert(_kv_only || mounted)

4. JobID: [6739515](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739515/)
- Failure Reason: Standalone test failed: osd-bluefs-volume-ops.sh
- src/os/bluestore/BlueStore.cc: 6684: FAILED ceph_assert(fm && fm->is_null_manager())

**Dead Jobs**
1. JobIDs: [6739268](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739268), [6739347](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739347), [6739425](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739425)
- Existing tracker: https://tracker.ceph.com/issues/53680

2. JobID: [6739311](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739311)
- Existing tracker: https://tracker.ceph.com/issues/51076

3. JobID: [6739505](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739505)
- Existing tracker: https://tracker.ceph.com/issues/52320

4. JobID: [6739563](http://pulpito.front.sepia.ceph.com/yuriw-2022-03-16_20:38:07-rados-wip-yuri3-testing-2022-03-16-1030-distro-default-smithi/6739563)
- No logs as test probably failed during init.

**Comment by sseshasa** [comment]
> I'd be surprised if any of those failures are a result of this PR. Let me know if they are.

@jtlayton None of them are related. And that's why they are all listed under "unrelated". I also double checked that the newly introduced test passed in the run. So it's all good.

**Pull request merged**